File: app/src/main/java/com/example/movies/MainViewModel.java
package com.example.movies;

import android.app.Application;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import java.util.List;

import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.annotations.SchedulerSupport;
import io.reactivex.rxjava3.disposables.CompositeDisposable;
import io.reactivex.rxjava3.disposables.Disposable;
import io.reactivex.rxjava3.schedulers.Schedulers;

public class MainViewModel extends AndroidViewModel {

    private static final String TAG = "MainViewModel";
    private final MutableLiveData<List<Movie>> movies = new MutableLiveData<List<Movie>>();
    private CompositeDisposable compositeDisposable = new CompositeDisposable();
    private int page = 1;

    public MainViewModel(@NonNull Application application) {
        super(application);
    }

    public LiveData<List<Movie>> getMovies() {
        return movies;
    }

    public void loadMovies() {
        Disposable disposable = ApiFactory.apiService.loadMovies(page)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(movieResponse -> {
                    List<Movie> movieListAdded = movies.getValue();
                    if(movieListAdded != null) {
                        movieListAdded.addAll(movieResponse.getMovies());
                        movies.setValue(movieListAdded);
                    } else {
                        movies.setValue(movieResponse.getMovies());
                    }
                    page++;
                }, error -> Log.d(TAG, "loadMovies: " + error.toString()));
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        compositeDisposable.dispose();
    }
}

File: app/src/main/java/com/example/movies/MainViewModel.kt
package com.example.movies

import android.util.Log
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.example.movies.api.ApiFactory
import com.example.movies.pojo.Movie
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.disposables.CompositeDisposable
import io.reactivex.rxjava3.schedulers.Schedulers

class MainViewModel : ViewModel() {

    private val compositeDisposable = CompositeDisposable()
    private val movies = MutableLiveData<List<Movie>>()
    private val isLoading = MutableLiveData<Boolean>()
    val getMovies: LiveData<List<Movie>> get() = movies
    val getIsLoading: LiveData<Boolean> get() = isLoading
    private var page = 1

    init {
        loadMovies()
    }

    fun loadMovies() {
        isLoading.value?.let { if (it) return }
        val disposable = ApiFactory.apiService.loadMovie(page)
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .doOnSubscribe { isLoading.value = true }
            .doAfterSuccess { isLoading.value = false }
            .subscribe({
                val movieListAdded = movies.value?.toMutableList()
                if (movieListAdded != null) {
                    movieListAdded.addAll(it.movies)
                    movies.value = movieListAdded
                } else {
                    movies.value = it.movies
                }
                page++
            }, { Log.e("MainViewModel", "loadMovies: $it") })
        compositeDisposable.add(disposable)
    }

    override fun onCleared() {
        super.onCleared()
        compositeDisposable.dispose()
    }
}